## Analisis exploratorio

### Librerias

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time
from sklearn.preprocessing import OneHotEncoder, MaxAbsScaler
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import sklearn.metrics as metrics
from sklearn.metrics import roc_auc_score
import lightgbm as lgb
from lightgbm import LGBMClassifier
from catboost import Pool, CatBoostClassifier
import optuna

ModuleNotFoundError: ignored

### Carga de datos

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
path = '/content/drive/MyDrive/TripleTen Projects/Proyecto Final/final_provider/'

def load_data(path,file_name):
    """
    Con esta funcion realizaremos la carga de datos, se realizara desde forma local o desde la plataforma tripleten, argumentos:
    path: ruta del archivo local
    file_name: nombre del archivo a cargar con su extension en str
    """
    try:
        df = pd.read_csv(path+file_name)
        return df
    except:
        df = pd.read_csv('/datasets/final_provider/'+file_name)
        return df

In [5]:
# Cargamos los datos
contract = load_data(path,'contract.csv')
internet = load_data(path,'internet.csv')
personal = load_data(path,'personal.csv')
phone = load_data(path,'phone.csv')

### Analisis exploratorio de datos

In [ ]:
def explore(data):
    """
    Con esta funcion podemos hacer una visualización de datos de forma rapida, con la cual podemos observar:
    - Los datos contenidos en el dataset
    - El tipo de los datos
    - Valores ausentes
    - Valores duplicados
    - Descripcion de las variables numericas en caso de contenerlas
    """
    display(data.head(10))
    print('Tipo de datos')
    print(data.info())
    print()
    print('Valores ausentes')
    print(data.isna().sum())
    print()
    print('Valores duplicados')
    print(data.duplicated().sum())
    print()
    print('Valores de variables númericas')
    print(data.describe())

#### Conjunto "contract"

In [ ]:
explore(contract)

customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   
5  9305-CDSKC  2019-03-01  2019-11-01 00:00:00  Month-to-month   
6  1452-KIOVK  2018-04-01                   No  Month-to-month   
7  6713-OKOMC  2019-04-01                   No  Month-to-month   
8  7892-POOKP  2017-07-01  2019-11-01 00:00:00  Month-to-month   
9  6388-TABGU  2014-12-01                   No        One year   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65  
5              Yes           Electronic check           99.65        820.5  
6              Yes    Credit card (automatic)           89.10       1949.4  
7               No               Mailed check           29.75        301.9  
8              Yes           Electronic check          104.80      3046.05  
9               No  Bank transfer (automatic)           56.15      3487.95

Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB
None

Valores ausentes
customerID          0
BeginDate           0
EndDate             0
Type                0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
dtype: int64

Valores duplicados
0

Valores de variables númericas
       MonthlyCharges
count     7043.000000
mean        64.761692
std         30.090047
mi

En este conjunto podemos ver:
* **"customerID"** - el ID del cliente
* **"BeginDate"** - su inicio de los servicios de la compañia
* **"EndDate"** - la fecha de terminacion del contrato si es que tiene
* **"Type"** - el tipo de contrato que tiene
* **"PaperlessBilling"** - si su facturacion es impresa o no
* **"PaymentMethod"** - metodo de pago
* **"MonthlyCharges"** - cargo mensual del pago
* **"TotalCharges"** - cargo total  que tendra que pagar

Podemos observar de entrada tenemos 7043 filas de datos, y los tipos de datos no corresponden a su tipo como las fechas y algunas variables numericas, los cuales deben ser corregidos.

En cuanto a las variables numericas podemos observar a simple que vista que no tenemos valores anormales, por lo que podemos considerarlos "correctos".

En cuanto a las variables categoricas veremos mas adelante si nos pueden ser de utilidad.

No contamos con valores ausentes o valores duplicados

#### Conjunto "internet"

In [ ]:
explore(internet)

customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0  7590-VHVEG             DSL             No          Yes               No   
1  5575-GNVDE             DSL            Yes           No              Yes   
2  3668-QPYBK             DSL            Yes          Yes               No   
3  7795-CFOCW             DSL            Yes           No              Yes   
4  9237-HQITU     Fiber optic             No           No               No   
5  9305-CDSKC     Fiber optic             No           No              Yes   
6  1452-KIOVK     Fiber optic             No          Yes               No   
7  6713-OKOMC             DSL            Yes           No               No   
8  7892-POOKP     Fiber optic             No           No              Yes   
9  6388-TABGU             DSL            Yes          Yes               No   

  TechSupport StreamingTV StreamingMovies  
0          No          No              No  
1          No          No              No  
2          No          No              No  
3         Yes          No              No  
4          No          No              No  
5          No         Yes             Yes  
6          No         Yes              No  
7          No          No              No  
8         Yes         Yes             Yes  
9          No          No              No

Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB
None

Valores ausentes
customerID          0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
dtype: int64

Valores duplicados
0

Valores de variables númericas
        customerID InternetService OnlineSecurity OnlineBackup  \
count         5517            5517           5

En este conjunto podemos ver:
* **"customerID"** - el ID del cliente
* **"InternetService"**	- Tipo de internet
* **"OnlineSecurity"** - Si tiene seguridad en linea
* **"OnlineBackup** - Si tiene respaldo en linea
* **"DeviceProtection** - Si tiene proteccion de dispositivo
* **"TechSupport** - Si tiene soporte de tecnologia
* **"StreamingTV** - Si tiene TV por internet
* **"StreamingMovies** - Si tiene peliculas por internet

Podemos ver que solo contiene 5517 filas en comparacion con los contactos que tiene 7043, por lo que podemos ver que hay clientes de los cuales no tenemos informacion.

Los tipos de datos son correctos al ser tipo object por contener solo palabras.

Contamos solo con variables categoricas, la mayoria con solo Yes/No como respuesta.

No contamos con valores ausentes o valores duplicados

#### Conjunto "personal"

In [ ]:
explore(personal)

customerID  gender  SeniorCitizen Partner Dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No
5  9305-CDSKC  Female              0      No         No
6  1452-KIOVK    Male              0      No        Yes
7  6713-OKOMC  Female              0      No         No
8  7892-POOKP  Female              0     Yes         No
9  6388-TABGU    Male              0      No        Yes

Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB
None

Valores ausentes
customerID       0
gender           0
SeniorCitizen    0
Partner          0
Dependents       0
dtype: int64

Valores duplicados
0

Valores de variables númericas
       SeniorCitizen
count    7043.000000
mean        0.162147
std         0.368612
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000


En este conjunto podemos ver:

* **"customerID"** - el ID del cliente
* **"gender"**	- Genero del cliente
* **"SeniorCitizen"** - Si es jubilado
* **"Partner** - Si tiene pareja
* **"Dependents** - Si tiene dependientes

Podemos ver que contiene 7043 filas,mismas filas que el conjunto contactos, por lo que podemos pensar que todos los datos estan completos.

Contamos solo con variables categoricas.

Los tipos de datos son correctos.

No contamos con valores ausentes o valores duplicados

#### Conjunto "phone"

In [ ]:
explore(phone)

customerID MultipleLines
0  5575-GNVDE            No
1  3668-QPYBK            No
2  9237-HQITU            No
3  9305-CDSKC           Yes
4  1452-KIOVK           Yes
5  7892-POOKP           Yes
6  6388-TABGU            No
7  9763-GRSKD            No
8  7469-LKBCI            No
9  8091-TTVAX           Yes

Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB
None

Valores ausentes
customerID       0
MultipleLines    0
dtype: int64

Valores duplicados
0

Valores de variables númericas
        customerID MultipleLines
count         6361          6361
unique        6361             2
top     5575-GNVDE            No
freq             1          3390


En este conjunto podemos ver:

* **"customerID"** - el ID del cliente
* **"MultipleLines"**	- Si tiene multiples lineas

Podemos ver que contiene 6361 filas, las cuales son menores a las 7043 del conjunto de contactos, por lo que podemos decir que no tenemos los datos completos.

Contamos solo con variables categoricas.

Los tipos de datos son correctos.

No contamos con valores ausentes o valores duplicados

## Preparación y correcion de datos

### Correcion de tipos de datos

Convertiremos los titulos de columnas a minusculas para un manejo mas facil

In [ ]:
def lower_col(data):
    """
    Convierte los titulos de las columnas a minusculas, argumentos:
    data: dataset a trabajar
    """
    new_names = []
    for x in data.columns:
        new_names.append(x.lower())
    data.columns = new_names

In [ ]:
# Crearemos una lista con nuestro 4 datasets en caso que requiramos hacer la misma instruccion en los 4
datasets = [contract,internet,personal,phone]

# Usamos la funcion "lower_col" en los 4 datasets
for x in datasets:
    lower_col(x)

In [ ]:
for x in datasets:
    display(x.head())

customerid   begindate              enddate            type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  paperlessbilling              paymentmethod  monthlycharges totalcharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65

customerid internetservice onlinesecurity onlinebackup deviceprotection  \
0  7590-VHVEG             DSL             No          Yes               No   
1  5575-GNVDE             DSL            Yes           No              Yes   
2  3668-QPYBK             DSL            Yes          Yes               No   
3  7795-CFOCW             DSL            Yes           No              Yes   
4  9237-HQITU     Fiber optic             No           No               No   

  techsupport streamingtv streamingmovies  
0          No          No              No  
1          No          No              No  
2          No          No              No  
3         Yes          No              No  
4          No          No              No

customerid  gender  seniorcitizen partner dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No

customerid multiplelines
0  5575-GNVDE            No
1  3668-QPYBK            No
2  9237-HQITU            No
3  9305-CDSKC           Yes
4  1452-KIOVK           Yes

Nuestros datasets ya se encuentran con los titulos de columnas en minusculas

Convertiremos los tipos de datos correspondientes de los 4 datasets que tenemos

In [ ]:
# Cambiamos el tipo de dato de fecha
contract['begindate'] = pd.to_datetime(contract['begindate'], format='%Y-%m-%d')

# La variables o columna 'enddate' no la cambiaremos ya que solo la usaremos como referencia para nuestro objetivo

# Convertimos a tipo de datos numericos
contract['totalcharges'] = pd.to_numeric(contract['totalcharges'],errors='coerce')

In [ ]:
# Revisamos los datos con valores ausentes
print('Datos ausentes')
print(contract['totalcharges'].isna().sum())
print()
print('Porcentraje de datos ausentes')
print(round((contract['totalcharges'].isna().sum()/len(contract))*100,2),'%')

Datos ausentes
11

Porcentraje de datos ausentes
0.16 %


No representa mayor detalle ya que es insignificante los valores ausente, los reemplazaremos por el promedio

In [ ]:
contract['totalcharges'].fillna(contract['totalcharges'].mean(),inplace=True)

### Creacion de columna objetivo

Crearemos nuestra columna con el valor "objetivo"

In [ ]:
# Creamos nuestra funcion de clasificacion de objetivo
def objetivo(row):
    if row['enddate'] == 'No':
        return 0
    else:
        return 1

# Aplicamos nuestra funcion para crear nuestra nueva columna objetivo
contract['churn'] = contract.apply(objetivo,axis=1)

In [ ]:
contract.head(10)

customerid  begindate              enddate            type  \
0  7590-VHVEG 2020-01-01                   No  Month-to-month   
1  5575-GNVDE 2017-04-01                   No        One year   
2  3668-QPYBK 2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW 2016-05-01                   No        One year   
4  9237-HQITU 2019-09-01  2019-11-01 00:00:00  Month-to-month   
5  9305-CDSKC 2019-03-01  2019-11-01 00:00:00  Month-to-month   
6  1452-KIOVK 2018-04-01                   No  Month-to-month   
7  6713-OKOMC 2019-04-01                   No  Month-to-month   
8  7892-POOKP 2017-07-01  2019-11-01 00:00:00  Month-to-month   
9  6388-TABGU 2014-12-01                   No        One year   

  paperlessbilling              paymentmethod  monthlycharges  totalcharges  \
0              Yes           Electronic check           29.85         29.85   
1               No               Mailed check           56.95       1889.50   
2              Yes               Mailed check           53.85        108.15   
3               No  Bank transfer (automatic)           42.30       1840.75   
4              Yes           Electronic check           70.70        151.65   
5              Yes           Electronic check           99.65        820.50   
6              Yes    Credit card (automatic)           89.10       1949.40   
7               No               Mailed check           29.75        301.90   
8              Yes           Electronic check          104.80       3046.05   
9               No  Bank transfer (automatic)           56.15       3487.95   

   churn  
0      0  
1      0  
2      1  
3      0  
4      1  
5      1  
6      0  
7      0  
8      1  
9      0

### Creacion de columna/variable dias de servicio

Crearemos nuestra columna con el valor de diferencia de dias, para esto crearemos una nueva columna referida a "enddate", en la cual si el valor es no, tomaremos como valida la ultima fecha registrada que es 1 de febrero de 2020 ó 2020-02-01

In [ ]:
# Creamos nuestra funcion de clasificacion de objetivo
def timetoint(row):
    if row['enddate'] == 'No':
        return '2020-02-01 00:00:00'
    else:
        return row['enddate']

# Aplicamos nuestra funcion para crear nuestra nueva columna objetivo
contract['enddate'] = contract.apply(timetoint,axis=1)

In [ ]:
# Cambiamos el tipo de dato de fecha
contract['enddate'] = pd.to_datetime(contract['enddate'], format='%Y-%m-%d')

In [ ]:
# Creamos nuestra variable
contract['diff_days'] = contract['enddate'] - contract['begindate']

In [ ]:
contract['diff_days'] = contract['diff_days'].dt.days.astype('int16')

In [ ]:
contract.head()

customerid  begindate    enddate            type paperlessbilling  \
0  7590-VHVEG 2020-01-01 2020-02-01  Month-to-month              Yes   
1  5575-GNVDE 2017-04-01 2020-02-01        One year               No   
2  3668-QPYBK 2019-10-01 2019-12-01  Month-to-month              Yes   
3  7795-CFOCW 2016-05-01 2020-02-01        One year               No   
4  9237-HQITU 2019-09-01 2019-11-01  Month-to-month              Yes   

               paymentmethod  monthlycharges  totalcharges  churn  diff_days  
0           Electronic check           29.85         29.85      0         31  
1               Mailed check           56.95       1889.50      0       1036  
2               Mailed check           53.85        108.15      1         61  
3  Bank transfer (automatic)           42.30       1840.75      0       1371  
4           Electronic check           70.70        151.65      1         61

### Union de conjuntos de datos

Para unir nuestros datos en un solo conjunto de datos usaremos como referencia la columna customerid, sabemos que no tenemos todos los datos de cada cliente en cada conjunto de datos, sin embargo, de momento no es problema alguno, continuaremos con la union de los diferentes conjuntos de datos

In [ ]:
# Crearemos nuestro conjunto de datos, usaremos "how='outer'" para verificar que efectivamente solo tengamos los 7043 datos de
# clientes unicos
df = pd.merge(contract,internet,how='outer',on='customerid')
df = pd.merge(df,personal,how='outer',on='customerid')
df = pd.merge(df,phone,how='outer',on='customerid')

In [ ]:
display(df.head())
print(df.info())

customerid  begindate    enddate            type paperlessbilling  \
0  7590-VHVEG 2020-01-01 2020-02-01  Month-to-month              Yes   
1  5575-GNVDE 2017-04-01 2020-02-01        One year               No   
2  3668-QPYBK 2019-10-01 2019-12-01  Month-to-month              Yes   
3  7795-CFOCW 2016-05-01 2020-02-01        One year               No   
4  9237-HQITU 2019-09-01 2019-11-01  Month-to-month              Yes   

               paymentmethod  monthlycharges  totalcharges  churn  diff_days  \
0           Electronic check           29.85         29.85      0         31   
1               Mailed check           56.95       1889.50      0       1036   
2               Mailed check           53.85        108.15      1         61   
3  Bank transfer (automatic)           42.30       1840.75      0       1371   
4           Electronic check           70.70        151.65      1         61   

   ... onlinebackup deviceprotection techsupport streamingtv streamingmovies  \
0  ...          Yes               No          No          No              No   
1  ...           No              Yes          No          No              No   
2  ...          Yes               No          No          No              No   
3  ...           No              Yes         Yes          No              No   
4  ...           No               No          No          No              No   

   gender seniorcitizen partner  dependents multiplelines  
0  Female             0     Yes          No           NaN  
1    Male             0      No          No            No  
2    Male             0      No          No            No  
3    Male             0      No          No           NaN  
4  Female             0      No          No            No  

[5 rows x 22 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customerid        7043 non-null   object        
 1   begindate         7043 non-null   datetime64[ns]
 2   enddate           7043 non-null   datetime64[ns]
 3   type              7043 non-null   object        
 4   paperlessbilling  7043 non-null   object        
 5   paymentmethod     7043 non-null   object        
 6   monthlycharges    7043 non-null   float64       
 7   totalcharges      7043 non-null   float64       
 8   churn             7043 non-null   int64         
 9   diff_days         7043 non-null   int16         
 10  internetservice   5517 non-null   object        
 11  onlinesecurity    5517 non-null   object        
 12  onlinebackup      5517 non-null   object        
 13  deviceprotection  5517 non-null   object        
 14  techsupport       5517 n

Con esto podemos validar que solo tenemos 7043 registros unicos de clientes de acuerdo a "customerid".

Ya tenemos nuestros 4 conjuntos de datos en una sola tabla, recordemos que tenemos valores que desconocemos en cuertas columnas, no les haremos caso por el momento.

En nuetro conjunto final tenemos 20 columnas, sin embargo, para fines de nuestro objetivo final y entrenamiento de modelo hay ciertas columnas que no son requeridas y/o no aportarian valor alguno a nuestro modelo como son: customerid,begindate,enddate; por lo que las eliminaremos de una vez.

In [ ]:
# Eliminamos las columnas mencionadas
df.drop(['customerid','begindate','enddate'],axis=1,inplace=True)

In [ ]:
display(df.head())
print(df.info())

type paperlessbilling              paymentmethod  monthlycharges  \
0  Month-to-month              Yes           Electronic check           29.85   
1        One year               No               Mailed check           56.95   
2  Month-to-month              Yes               Mailed check           53.85   
3        One year               No  Bank transfer (automatic)           42.30   
4  Month-to-month              Yes           Electronic check           70.70   

   totalcharges  churn  diff_days internetservice onlinesecurity onlinebackup  \
0         29.85      0         31             DSL             No          Yes   
1       1889.50      0       1036             DSL            Yes           No   
2        108.15      1         61             DSL            Yes          Yes   
3       1840.75      0       1371             DSL            Yes           No   
4        151.65      1         61     Fiber optic             No           No   

  deviceprotection techsupport streamingtv streamingmovies  gender  \
0               No          No          No              No  Female   
1              Yes          No          No              No    Male   
2               No          No          No              No    Male   
3              Yes         Yes          No              No    Male   
4               No          No          No              No  Female   

   seniorcitizen partner dependents multiplelines  
0              0     Yes         No           NaN  
1              0      No         No            No  
2              0      No         No            No  
3              0      No         No           NaN  
4              0      No         No            No

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   type              7043 non-null   object 
 1   paperlessbilling  7043 non-null   object 
 2   paymentmethod     7043 non-null   object 
 3   monthlycharges    7043 non-null   float64
 4   totalcharges      7043 non-null   float64
 5   churn             7043 non-null   int64  
 6   diff_days         7043 non-null   int16  
 7   internetservice   5517 non-null   object 
 8   onlinesecurity    5517 non-null   object 
 9   onlinebackup      5517 non-null   object 
 10  deviceprotection  5517 non-null   object 
 11  techsupport       5517 non-null   object 
 12  streamingtv       5517 non-null   object 
 13  streamingmovies   5517 non-null   object 
 14  gender            7043 non-null   object 
 15  seniorcitizen     7043 non-null   int64  
 16  partner           7043 non-null   object 


Ya tenemos nuestro conjunto un 50% listo para poder usarlos en un modelo de machine learning.

Hablando en terminos de machine learning tenemos 2 varibles numericas y 15 variables categoricas y 1 variable objetivo:
- Vamos a realizar la codificacion OneHotEncoding para nuestras variables categoricas que en este caso serian 14 ya que la columna "seniorcitizen" ya esta clasificada
- Vamos a realizar un escalamiento de datos para nuestras variables numericas: monthlycharges, totalcharges
- Dejaremos nuestra columna "churn" tal cual, ya que no requiere modificaciones y es nuestra columna objetivo

In [ ]:
# Definimos en listas nuestras variables
categorical = ['type', 'paymentmethod', 'internetservice', 'onlinesecurity',
       'onlinebackup', 'deviceprotection', 'techsupport', 'streamingtv',
       'streamingmovies', 'gender', 'seniorcitizen', 'partner', 'dependents',
       'multiplelines','paperlessbilling', ]
numerical = ['monthlycharges',
       'totalcharges','diff_days']

In [ ]:
# Revisaremos de forma rapida los valores unicos de nuestras variables categoricas para verificar que no haya datos erroneos
for x in categorical:
    print('Valores unicos de:',x)
    print(df[x].sort_values().unique())
    print()

Valores unicos de: type
['Month-to-month' 'One year' 'Two year']

Valores unicos de: paymentmethod
['Bank transfer (automatic)' 'Credit card (automatic)' 'Electronic check'
 'Mailed check']

Valores unicos de: internetservice
['DSL' 'Fiber optic' nan]

Valores unicos de: onlinesecurity
['No' 'Yes' nan]

Valores unicos de: onlinebackup
['No' 'Yes' nan]

Valores unicos de: deviceprotection
['No' 'Yes' nan]

Valores unicos de: techsupport
['No' 'Yes' nan]

Valores unicos de: streamingtv
['No' 'Yes' nan]

Valores unicos de: streamingmovies
['No' 'Yes' nan]

Valores unicos de: gender
['Female' 'Male']

Valores unicos de: seniorcitizen
[0 1]

Valores unicos de: partner
['No' 'Yes']

Valores unicos de: dependents
['No' 'Yes']

Valores unicos de: multiplelines
['No' 'Yes' nan]

Valores unicos de: paperlessbilling
['No' 'Yes']



Podemos ver que en la columna de "paymentmethod" tenemos 4 tipos de datos, sin embargo, podemos convertirlos a dos, ya que dos de los tipos de datos es de forma automatica y dos en pago con cheque, asi que los agruparemos

In [ ]:
# Creamos nuestra funcion de clasificacion de objetivo
def col_paymentmethod(row):
    if row['paymentmethod'] == 'Bank transfer (automatic)':
        return 'automatic'
    if row['paymentmethod'] == 'Credit card (automatic)':
        return 'automatic'
    else:
        return 'check'

# Aplicamos nuestra funcion para crear nuestra nueva columna objetivo
df['paymentmethod'] = df.apply(col_paymentmethod,axis=1)

# Comprobamos los valores unicos nuevamente
print(df['paymentmethod'].sort_values().unique())

['automatic' 'check']


Podemos ver que en la columna de "type" tenemos 3 tipos de datos, sin embargo, podemos convertirlos a dos, ya que dos de los tipos de datos son contratos de forma anual y uno de forma mensual, asi que agruparemos los de tipo anual

In [ ]:
# Creamos nuestra funcion de clasificacion de objetivo
def col_type(row):
    if row['type'] == 'Month-to-month':
        return 'monthly'
    else:
        return 'anual'

# Aplicamos nuestra funcion para crear nuestra nueva columna objetivo
df['type'] = df.apply(col_type,axis=1)

# Comprobamos los valores unicos nuevamente
print(df['type'].sort_values().unique())

['anual' 'monthly']


In [ ]:
df.fillna('unkn',inplace=True)

Con esto podemos ver los valores unicos de cada variable categorica, ahora, tenemos valores ausentes, podemos llenar estos valores ausentes con la palabra "unknown", sin embargo, al codificar nuestras variables automaticamente ese valor unknown representaria lo mismo que un valor "NaN", seria una etiqueta extra, por lo que para fines practicos dejaremos los valores ausentes ya que asi nos ahorramos tiempo en un proceso que el resultado seria el mismo, asi que manos a la obra.

### Codificación OneHotEncoding

In [ ]:
# Crearemos nuestro codificador y convertiremos los datos
ohe = OneHotEncoder(sparse = False, drop='if_binary')
ohe_data = pd.DataFrame(ohe.fit_transform(df[categorical]), columns = ohe.get_feature_names_out())

In [ ]:
ohe_data

type_monthly  paymentmethod_check  internetservice_DSL  \
0              1.0                  1.0                  1.0   
1              0.0                  1.0                  1.0   
2              1.0                  1.0                  1.0   
3              0.0                  0.0                  1.0   
4              1.0                  1.0                  0.0   
...            ...                  ...                  ...   
7038           0.0                  1.0                  1.0   
7039           0.0                  0.0                  0.0   
7040           1.0                  1.0                  1.0   
7041           1.0                  1.0                  0.0   
7042           0.0                  0.0                  0.0   

      internetservice_Fiber optic  internetservice_unkn  onlinesecurity_No  \
0                             0.0                   0.0                1.0   
1                             0.0                   0.0                0.0   
2                             0.0                   0.0                0.0   
3                             0.0                   0.0                0.0   
4                             1.0                   0.0                1.0   
...                           ...                   ...                ...   
7038                          0.0                   0.0                0.0   
7039                          1.0                   0.0                1.0   
7040                          0.0                   0.0                0.0   
7041                          1.0                   0.0                1.0   
7042                          1.0                   0.0                0.0   

      onlinesecurity_Yes  onlinesecurity_unkn  onlinebackup_No  \
0                    0.0                  0.0              0.0   
1                    1.0                  0.0              1.0   
2                    1.0                  0.0              0.0   
3                    1.0                  0.0              1.0   
4                    0.0                  0.0              1.0   
...                  ...                  ...              ...   
7038                 1.0                  0.0              1.0   
7039                 0.0                  0.0              0.0   
7040                 1.0                  0.0              1.0   
7041                 0.0                  0.0              1.0   
7042                 1.0                  0.0              1.0   

      onlinebackup_Yes  ...  streamingmovies_Yes  streamingmovies_unkn  \
0                  1.0  ...                  0.0                   0.0   
1                  0.0  ...                  0.0                   0.0   
2                  1.0  ...                  0.0                   0.0   
3                  0.0  ...                  0.0                   0.0   
4                  0.0  ...                  0.0                   0.0   
...                ...  ...                  ...                   ...   
7038               0.0  ...                  1.0                   0.0   
7039               1.0  ...                  1.0                   0.0   
7040               0.0  ...                  0.0                   0.0   
7041               0.0  ...                  0.0                   0.0   
7042               0.0  ...                  1.0                   0.0   

      gender_Male  seniorcitizen_1  partner_Yes  dependents_Yes  \
0             0.0              0.0          1.0             0.0   
1             1.0              0.0          0.0             0.0   
2             1.0              0.0          0.0             0.0   
3             1.0              0.0          0.0             0.0   
4             0.0              0.0          0.0             0.0   
...           ...              ...          ...             ...   
7038          1.0              0.0          1.0             1.0   
7039          0.0              0.0          1.0             1.0   
7040          0

### Escalado de datos

In [ ]:
# Creamos nuestro escalador de datos y convertiremos los datos
scaler = MaxAbsScaler()
scal_data = pd.DataFrame(scaler.fit_transform(df[numerical]),columns=df[numerical].columns)

In [ ]:
scal_data

monthlycharges  totalcharges  diff_days
0           0.251368      0.003437   0.014149
1           0.479579      0.217564   0.472843
2           0.453474      0.012453   0.027841
3           0.356211      0.211951   0.625742
4           0.595368      0.017462   0.027841
...              ...           ...        ...
7038        0.714105      0.229194   0.333181
7039        0.869053      0.847792   1.000000
7040        0.249263      0.039892   0.153811
7041        0.626526      0.035303   0.056139
7042        0.889684      0.788101   0.917389

[7043 rows x 3 columns]

Ya tenemos nuestras variables con el escalamiento de datos, procederemos a juntar nuestros conjuntos de OHE y escalamiento

### Features y target

Ya tenemos nuestros datos listos para definirlos como features y target

####  Features

In [ ]:
features = pd.concat([ohe_data,scal_data],axis=1)

In [ ]:
features

type_monthly  paymentmethod_check  internetservice_DSL  \
0              1.0                  1.0                  1.0   
1              0.0                  1.0                  1.0   
2              1.0                  1.0                  1.0   
3              0.0                  0.0                  1.0   
4              1.0                  1.0                  0.0   
...            ...                  ...                  ...   
7038           0.0                  1.0                  1.0   
7039           0.0                  0.0                  0.0   
7040           1.0                  1.0                  1.0   
7041           1.0                  1.0                  0.0   
7042           0.0                  0.0                  0.0   

      internetservice_Fiber optic  internetservice_unkn  onlinesecurity_No  \
0                             0.0                   0.0                1.0   
1                             0.0                   0.0                0.0   
2                             0.0                   0.0                0.0   
3                             0.0                   0.0                0.0   
4                             1.0                   0.0                1.0   
...                           ...                   ...                ...   
7038                          0.0                   0.0                0.0   
7039                          1.0                   0.0                1.0   
7040                          0.0                   0.0                0.0   
7041                          1.0                   0.0                1.0   
7042                          1.0                   0.0                0.0   

      onlinesecurity_Yes  onlinesecurity_unkn  onlinebackup_No  \
0                    0.0                  0.0              0.0   
1                    1.0                  0.0              1.0   
2                    1.0                  0.0              0.0   
3                    1.0                  0.0              1.0   
4                    0.0                  0.0              1.0   
...                  ...                  ...              ...   
7038                 1.0                  0.0              1.0   
7039                 0.0                  0.0              0.0   
7040                 1.0                  0.0              1.0   
7041                 0.0                  0.0              1.0   
7042                 1.0                  0.0              1.0   

      onlinebackup_Yes  ...  seniorcitizen_1  partner_Yes  dependents_Yes  \
0                  1.0  ...              0.0          1.0             0.0   
1                  0.0  ...              0.0          0.0             0.0   
2                  1.0  ...              0.0          0.0             0.0   
3                  0.0  ...              0.0          0.0             0.0   
4                  0.0  ...              0.0          0.0             0.0   
...                ...  ...              ...          ...             ...   
7038               0.0  ...              0.0          1.0             1.0   
7039               1.0  ...              0.0          1.0             1.0   
7040               0.0  ...              0.0          1.0             1.0   
7041               0.0  ...              1.0          1.0             0.0   
7042               0.0  ...              0.0          0.0             0.0   

      multiplelines_No  multiplelines_Yes  multiplelines_unkn  \
0                  0.0                0.0                 1.0   
1                  1.0                0.0                 0.0   
2                  1.0                0.0                 0.0   
3                  0.0                0.0                 1.0   
4                  1.0                0.0                 0.0   
...                ...                ...                 ...   
7038               0.0                1.0                 0.0   
7039               0.0                1.0                 0.0 

####  Target

In [ ]:
target = df['churn']

In [ ]:
target

0       0
1       0
2       1
3       0
4       1
       ..
7038    0
7039    0
7040    0
7041    1
7042    0
Name: churn, Length: 7043, dtype: int64

Ya tenemos nuestras features y target listos para entrenar nuestros modelos de clasificación, manos a la obra

## Modelado de datos

Crearemos un dataframe para ir comparando los diferentes modelos y sus resultados:

In [ ]:
global_results = {'tipo':[],
                  'modelo':[],
                  'AUC-ROC':[],
                  'tiempo_entrenamiento':[]}
global_results = pd.DataFrame(global_results)

Crearemos nuestros conjuntos de entrenamiento y prueba 4:1

In [ ]:
def add_results(train_result,valid_result,time,model,table):
    """Con esta funcion agregaremos los valores del nombre y tipo de conjunto, su valor de RECM, el tiempo de procesamiento/entrenamiento,
    y el tipo de modelo, con los siguientes argumentos
    train_result: resultado de RECM conjunto entrenamiento
    valid_result: resultado de RECM conjunto validacion
    time: resultado del tiempo de procesamiento
    model: tipo de modelo,
    table: tabla a la que se agregaran los resultados
    """
    row1 = pd.DataFrame({'tipo' : ['train'], 'modelo':[model],'AUC-ROC': [train_result],'tiempo_entrenamiento': [time]})
    row2 = pd.DataFrame({'tipo' : ['valid'], 'modelo':[model],'AUC-ROC': [valid_result],'tiempo_entrenamiento': [time]})
    table = pd.concat([table,row1,row2])
    return table.reset_index(drop=True)

### Separacion de datos train, test

In [ ]:
features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.20, random_state=92)

In [ ]:
# Verificamos el tamaño de nuestros conjuntos
features_train.shape, features_valid.shape, target_train.shape, target_valid.shape

((5634, 34), (1409, 34), (5634,), (1409,))

### Balanceo de clases

Algo que puede afectar una tarea de clasificacion es el desbalanceo de clases, veamos cual es la situacion actualy tomemos la decision de hacer un sobremuestreo o un submuestreo

In [ ]:
features_zeros =  features_train[target_train == 0]
features_ones = features_train[target_train == 1]
target_zeros = target_train[target_train == 0]
target_ones = target_train[target_train == 1]

print(features_zeros.shape)
print(features_ones.shape)
print(target_zeros.shape)
print(target_ones.shape)

(4133, 34)
(1501, 34)
(4133,)
(1501,)


In [ ]:
print('Balance entre clases')
print()
print('Porcentaje de "1"')
print(round(len(target_ones)/(len(target_ones)+len(target_zeros))*100,2),'%')
print()
print('Porcentaje de "0"')
print(round(len(target_zeros)/(len(target_ones)+len(target_zeros))*100,2),'%')

Balance entre clases

Porcentaje de "1"
26.64 %

Porcentaje de "0"
73.36 %


Podemos ver que tenemos un desbalance de clases 3:1 solo en el split de entrenamiento, tenemos entendido que un desbalance de clases en un modelo de clasificacion puedeafectar nuestros resultados finales, en este caso nuestro valor de AUC-ROC, hagamos un balanceo haciendo un sobremuestreo con los target = 1

In [ ]:
def upsample(features, target, repeat):
    """
    Con esta funcion haremos el sobremuestreo con tres argumentos:
    features: conjunto de caracteristicas
    target: serie de objetivos
    repeat: el numero de veces que queremos el sobremuestreo
    """
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345
    )

    return features_upsampled, target_upsampled

In [ ]:
features_train, target_train = upsample(features_train, target_train, 4)

In [ ]:
# Comprobamos una vez mas el tamaño de los conjuntos y el balance de clases
features_zeros =  features_train[target_train == 0]
features_ones = features_train[target_train == 1]
target_zeros = target_train[target_train == 0]
target_ones = target_train[target_train == 1]

print(features_zeros.shape)
print(features_ones.shape)
print(target_zeros.shape)
print(target_ones.shape)

(4133, 34)
(6004, 34)
(4133,)
(6004,)


### Logistic Regression

In [ ]:
start_time = time.time()
model = LogisticRegression(random_state=92)
param = {'solver':['liblinear'],
         'penalty':['l1','l2']}
gs = GridSearchCV(estimator = model,
                  param_grid = param,
                  scoring = 'roc_auc',
                  cv = 5)
gs.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejor modelo:',gs.best_estimator_)
train_pred = gs.best_estimator_.predict(features_train)
valid_pred = gs.best_estimator_.predict(features_valid)
# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

Tiempo de entrenamiento: 0.54 segundos
Mejor modelo: LogisticRegression(penalty='l1', random_state=92, solver='liblinear')
La puntuación auc-roc del conjunto de entrenamiento es: 0.7606218807592261
La puntuación auc-roc del conjunto de validación es: 0.7562857620181263


In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'LogisticRegression',global_results)

### Logistic Regression - Optuna

In [ ]:
def objective(trial):
    # Hiperparametros
    penalty = trial.suggest_categorical('penalty', ['none', 'l2',])
    solver = trial.suggest_categorical('solver', ['lbfgs', 'saga','newton-cg'])
    C = trial.suggest_float('C',1,100)
    max_iter = trial.suggest_int('max_iter',100,200)

    # Modelo
    model = LogisticRegression(random_state=92,penalty=penalty,solver=solver,C=C,max_iter=max_iter)
    model.fit(features_train,target_train)
    valid_pred = model.predict(features_valid)
    auc_roc = roc_auc_score(target_valid,valid_pred)
    return auc_roc

# Creacion de estudio
start_time = time.time()
study = optuna.create_study(direction='maximize')
study.optimize(objective,n_trials=200)
best_params = study.best_params
# Uso de mejores hiperparametros
model = LogisticRegression(random_state=92,penalty=best_params['penalty'],solver=best_params['solver'],
                           C=best_params['C'],max_iter=best_params['max_iter'])
model.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejores hiperparametros:',study.best_params)
train_pred = model.predict(features_train)
valid_pred = model.predict(features_valid)

# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

[I 2023-10-23 13:03:47,030] A new study created in memory with name: no-name-472a26f4-549f-4aea-a342-ef9d5b08576b
[I 2023-10-23 13:03:47,079] Trial 0 finished with value: 0.7572463768115942 and parameters: {'penalty': 'l2', 'solver': 'lbfgs', 'C': 97.7777432419157, 'max_iter': 151}. Best is trial 0 with value: 0.7572463768115942.
[I 2023-10-23 13:03:47,124] Trial 1 finished with value: 0.7572463768115942 and parameters: {'penalty': 'l2', 'solver': 'lbfgs', 'C': 64.34497489990679, 'max_iter': 121}. Best is trial 0 with value: 0.7572463768115942.
[I 2023-10-23 13:03:47,169] Trial 2 finished with value: 0.7563679885561542 and parameters: {'penalty': 'l2', 'solver': 'lbfgs', 'C': 33.02884189575636, 'max_iter': 174}. Best is trial 0 with value: 0.7572463768115942.
[I 2023-10-23 13:03:47,214] Trial 3 finished with value: 0.7563679885561542 and parameters: {'penalty': 'l2', 'solver': 'lbfgs', 'C': 30.61024100740888, 'max_iter': 159}. Best is trial 0 with value: 0.7572463768115942.
C:\Users\us

Tiempo de entrenamiento: 49.32 segundos
Mejores hiperparametros: {'penalty': 'l2', 'solver': 'lbfgs', 'C': 4.098784693354367, 'max_iter': 143}
La puntuación auc-roc del conjunto de entrenamiento es: 0.7597137032445342
La puntuación auc-roc del conjunto de validación es: 0.7586050724637683


In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'LogisticRegression-optuna',global_results)

### Decision Tree Classifier

In [ ]:
start_time = time.time()
model = DecisionTreeClassifier(random_state=92)
param = {'criterion':['gini', 'entropy'],
         'max_depth':[None,5,10,20,50],
         'min_samples_split':[2,3,4,5]
        }
gs = GridSearchCV(estimator = model,
                  param_grid = param,
                  scoring = 'roc_auc',
                  cv = 5)
gs.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejor modelo:',gs.best_estimator_)
train_pred = gs.best_estimator_.predict(features_train)
valid_pred = gs.best_estimator_.predict(features_valid)
# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

Tiempo de entrenamiento: 6.18 segundos
Mejor modelo: DecisionTreeClassifier(max_depth=20, min_samples_split=3, random_state=92)
La puntuación auc-roc del conjunto de entrenamiento es: 0.978706267762777
La puntuación auc-roc del conjunto de validación es: 0.7387532368541955


In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'DecisionTreeClassifier',global_results)

### Decision Tree Classifier - Optuna

In [ ]:
def objective(trial):
    # Hiperparametros
    criterion = trial.suggest_categorical('criterion', ['gini', 'entropy'])
    splitter = trial.suggest_categorical('splitter', ['random', 'best'])
    max_features = trial.suggest_categorical('max_features', ['sqrt', 'log2','auto'])
    max_depth = trial.suggest_int('max_depth',1,20)
    min_samples_split = trial.suggest_int('min_samples_split',2,20)
    min_samples_leaf = trial.suggest_int('min_samples_leaf',1,20)
    min_weight_fraction_leaf = trial.suggest_float('min_weight_fraction_leaf',0.0,0.5)

    # Modelo
    model = DecisionTreeClassifier(random_state=92,criterion=criterion,splitter=splitter,max_features=max_features,
                                   max_depth=max_depth,min_samples_split=min_samples_split,
                                   min_samples_leaf=min_samples_leaf,min_weight_fraction_leaf=min_weight_fraction_leaf)
    model.fit(features_train,target_train)
    valid_pred = model.predict(features_valid)
    auc_roc = roc_auc_score(target_valid,valid_pred)
    return auc_roc

# Creacion de estudio
start_time = time.time()
study = optuna.create_study(direction='maximize')
study.optimize(objective,n_trials=200)
best_params = study.best_params
# Uso de mejores hiperparametros
model = DecisionTreeClassifier(random_state=92,criterion=best_params['criterion'],splitter=best_params['splitter'],
                               max_features=best_params['max_features'],max_depth=best_params['max_depth'],
                               min_samples_split=best_params['min_samples_split'],min_samples_leaf=best_params['min_samples_leaf'],
                               min_weight_fraction_leaf=best_params['min_weight_fraction_leaf'])
model.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejores hiperparametros:',study.best_params)
train_pred = model.predict(features_train)
valid_pred = model.predict(features_valid)

# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

[I 2023-10-23 13:04:42,604] A new study created in memory with name: no-name-ea6e1e73-cfed-48b6-8307-d45d229a6647
[I 2023-10-23 13:04:42,613] Trial 0 finished with value: 0.740626174664829 and parameters: {'criterion': 'gini', 'splitter': 'random', 'max_features': 'sqrt', 'max_depth': 10, 'min_samples_split': 11, 'min_samples_leaf': 20, 'min_weight_fraction_leaf': 0.07462769037823247}. Best is trial 0 with value: 0.740626174664829.
[I 2023-10-23 13:04:42,620] Trial 1 finished with value: 0.6439812784529926 and parameters: {'criterion': 'entropy', 'splitter': 'best', 'max_features': 'sqrt', 'max_depth': 1, 'min_samples_split': 20, 'min_samples_leaf': 12, 'min_weight_fraction_leaf': 0.05341530658974686}. Best is trial 0 with value: 0.740626174664829.
[I 2023-10-23 13:04:42,627] Trial 2 finished with value: 0.5 and parameters: {'criterion': 'gini', 'splitter': 'random', 'max_features': 'log2', 'max_depth': 12, 'min_samples_split': 17, 'min_samples_leaf': 4, 'min_weight_fraction_leaf': 0.3

Tiempo de entrenamiento: 5.57 segundos
Mejores hiperparametros: {'criterion': 'entropy', 'splitter': 'random', 'max_features': 'sqrt', 'max_depth': 11, 'min_samples_split': 13, 'min_samples_leaf': 20, 'min_weight_fraction_leaf': 0.0003082308108589677}
La puntuación auc-roc del conjunto de entrenamiento es: 0.7609004272174062
La puntuación auc-roc del conjunto de validación es: 0.7551476423171699


In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'DecisionTreeClassifier - optuna',global_results)

### Random Forest Classifier

In [ ]:
start_time = time.time()
model = RandomForestClassifier(random_state=92)
param = {'n_estimators':[100,200],
         'max_depth':[None,5,10],
         'min_samples_split':[2,3,4,5]
        }
gs = GridSearchCV(estimator = model,
                  param_grid = param,
                  scoring = 'roc_auc',
                  cv = 5)
gs.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejor modelo:',gs.best_estimator_)
train_pred = gs.best_estimator_.predict(features_train)
valid_pred = gs.best_estimator_.predict(features_valid)
# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

Tiempo de entrenamiento: 72.63 segundos
Mejor modelo: RandomForestClassifier(n_estimators=200, random_state=92)
La puntuación auc-roc del conjunto de entrenamiento es: 0.9987902250181466
La puntuación auc-roc del conjunto de validación es: 0.7546033809464144


In [ ]:
gs.best_params_

{'max_depth': None, 'min_samples_split': 2, 'n_estimators': 200}

In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'RandomForestClassifier',global_results)

### Random Forest Classifier - Optuna

In [ ]:
def objective(trial):
    # Hiperparametros
    criterion = trial.suggest_categorical('criterion', ['gini', 'entropy'])
    n_estimators = trial.suggest_int('n_estimators',10,300)
    max_features = trial.suggest_categorical('max_features', ['sqrt', 'log2','auto'])
    max_depth = trial.suggest_int('max_depth',1,20)
    min_samples_split = trial.suggest_int('min_samples_split',2,20)
    min_samples_leaf = trial.suggest_int('min_samples_leaf',1,20)
    min_weight_fraction_leaf = trial.suggest_float('min_weight_fraction_leaf',0.0,0.5)

    # Modelo
    model = RandomForestClassifier(random_state=92,n_estimators=n_estimators,criterion=criterion,max_features=max_features,
                                   max_depth=max_depth,min_samples_split=min_samples_split,
                                   min_samples_leaf=min_samples_leaf,min_weight_fraction_leaf=min_weight_fraction_leaf)
    model.fit(features_train,target_train)
    valid_pred = model.predict(features_valid)
    auc_roc = roc_auc_score(target_valid,valid_pred)
    return auc_roc

# Creacion de estudio
start_time = time.time()
study = optuna.create_study(direction='maximize')
study.optimize(objective,n_trials=200)
best_params = study.best_params
# Uso de mejores hiperparametros
model = RandomForestClassifier(random_state=92,n_estimators=best_params['n_estimators'],criterion=best_params['criterion'],
                               max_features=best_params['max_features'],max_depth=best_params['max_depth'],
                               min_samples_split=best_params['min_samples_split'],min_samples_leaf=best_params['min_samples_leaf'],
                               min_weight_fraction_leaf=best_params['min_weight_fraction_leaf'])
model.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejores hiperparametros:',study.best_params)
train_pred = model.predict(features_train)
valid_pred = model.predict(features_valid)

# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

[I 2023-10-23 13:06:01,236] A new study created in memory with name: no-name-a46cb9ec-9be4-4092-adb2-c0a2ab1a3ab0
[I 2023-10-23 13:06:01,750] Trial 0 finished with value: 0.7501239923986134 and parameters: {'criterion': 'entropy', 'n_estimators': 163, 'max_features': 'auto', 'max_depth': 6, 'min_samples_split': 14, 'min_samples_leaf': 10, 'min_weight_fraction_leaf': 0.07258304591304493}. Best is trial 0 with value: 0.7501239923986134.
[I 2023-10-23 13:06:02,017] Trial 1 finished with value: 0.7255500041765861 and parameters: {'criterion': 'gini', 'n_estimators': 120, 'max_features': 'auto', 'max_depth': 14, 'min_samples_split': 11, 'min_samples_leaf': 16, 'min_weight_fraction_leaf': 0.23386217198624631}. Best is trial 0 with value: 0.7501239923986134.
[I 2023-10-23 13:06:02,184] Trial 2 finished with value: 0.6962773566386835 and parameters: {'criterion': 'gini', 'n_estimators': 83, 'max_features': 'auto', 'max_depth': 2, 'min_samples_split': 4, 'min_samples_leaf': 19, 'min_weight_frac

Tiempo de entrenamiento: 112.63 segundos
Mejores hiperparametros: {'criterion': 'gini', 'n_estimators': 136, 'max_features': 'auto', 'max_depth': 12, 'min_samples_split': 2, 'min_samples_leaf': 16, 'min_weight_fraction_leaf': 0.0008443219060412081}
La puntuación auc-roc del conjunto de entrenamiento es: 0.8117312871344904
La puntuación auc-roc del conjunto de validación es: 0.7804525330994445


In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'RandomForestClassifier - optuna',global_results)

### LGBM Classifier

In [ ]:
start_time = time.time()
model = LGBMClassifier()
model.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejor modelo:',gs.best_estimator_)
train_pred = model.predict(features_train)
valid_pred = model.predict(features_valid)
# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 6004, number of negative: 4133
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000431 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 728
[LightGBM] [Info] Number of data points in the train set: 10137, number of used features: 34
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.592286 -> initscore=0.373422
[LightGBM] [Info] Start training from score 0.373422
Tiempo de entrenamiento: 0.2 segundos
Mejor modelo: RandomForestClassifier(n_estimators=200, random_state=92)
La puntuación auc-roc del conjunto de entrenamiento es: 0.9197187357795021
La puntuación auc-roc del conjunto de validación es: 0.8364148707346615


C:\Users\user\anaconda3\lib\site-packages\joblib\externals\loky\backend\context.py:150: UserWarning: Could not find the number of physical cores for the following reason:
found 0 physical cores < 1
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "C:\Users\user\anaconda3\lib\site-packages\joblib\externals\loky\backend\context.py", line 245, in _count_physical_cores
    raise ValueError(


In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'LGBMClassifier',global_results)

### Catboost Classifier

In [ ]:
start_time = time.time()
model = CatBoostClassifier(eval_metric='AUC')
model.fit(features_train,target_train)
end_time = time.time()
training_time = round(end_time - start_time,2)
print('Tiempo de entrenamiento:', training_time,'segundos')
print('Mejor modelo:',gs.best_estimator_)
train_pred = model.predict(features_train)
valid_pred = model.predict(features_valid)
# AUC-Score
auc_train = roc_auc_score(target_train,train_pred)
print('La puntuación auc-roc del conjunto de entrenamiento es:',auc_train)
auc_valid = roc_auc_score(target_valid,valid_pred)
print('La puntuación auc-roc del conjunto de validación es:',auc_valid)

Learning rate set to 0.027699
0:	total: 149ms	remaining: 2m 29s
1:	total: 157ms	remaining: 1m 18s
2:	total: 164ms	remaining: 54.6s
3:	total: 172ms	remaining: 42.9s
4:	total: 179ms	remaining: 35.7s
5:	total: 188ms	remaining: 31.1s
6:	total: 195ms	remaining: 27.7s
7:	total: 203ms	remaining: 25.2s
8:	total: 211ms	remaining: 23.2s
9:	total: 219ms	remaining: 21.7s
10:	total: 226ms	remaining: 20.4s
11:	total: 234ms	remaining: 19.2s
12:	total: 241ms	remaining: 18.3s
13:	total: 247ms	remaining: 17.4s
14:	total: 253ms	remaining: 16.6s
15:	total: 260ms	remaining: 16s
16:	total: 267ms	remaining: 15.4s
17:	total: 273ms	remaining: 14.9s
18:	total: 279ms	remaining: 14.4s
19:	total: 286ms	remaining: 14s
20:	total: 292ms	remaining: 13.6s
21:	total: 298ms	remaining: 13.3s
22:	total: 305ms	remaining: 12.9s
23:	total: 311ms	remaining: 12.6s
24:	total: 318ms	remaining: 12.4s
25:	total: 324ms	remaining: 12.1s
26:	total: 331ms	remaining: 11.9s
27:	total: 337ms	remaining: 11.7s
28:	total: 344ms	remaining: 11

In [ ]:
global_results = add_results(auc_train,auc_valid,training_time,'CatBoostClassifier',global_results)

### Resultados

In [ ]:
global_results.query('tipo == "valid"').sort_values(by='AUC-ROC',ascending=False)

tipo                           modelo   AUC-ROC  tiempo_entrenamiento
13  valid                   LGBMClassifier  0.836415                  0.20
15  valid               CatBoostClassifier  0.825216                  5.71
11  valid  RandomForestClassifier - optuna  0.780453                112.63
3   valid        LogisticRegression-optuna  0.758605                 49.32
1   valid               LogisticRegression  0.756286                  0.54
7   valid  DecisionTreeClassifier - optuna  0.755148                  5.57
9   valid           RandomForestClassifier  0.754603                 72.63
5   valid           DecisionTreeClassifier  0.738753                  6.18

## Conclusión

Para llegar a una solucion de este proyecto primero corregimos los datos de ciertas columnas como lo son fechas y datos numericos al convertirlos al tipo correspondiente de datos.

Creamos dos variables:
1. La variable objetivo que en este caso es "churn" respecto a si ya habian finalizado su servicio o no.
2. La variable de "diff_days", que es la cantidad de dias que un usuario a utlizado el servicio.

Unimos los 4 conjuntos de datos para obtener una sola tabla.

Revisamos los diferentes tipos de variables categoricas en cada columna para verificar que no existieran valores que podamos agrupar, siendo este el caso los agrupamos para tener una homologacion de datos.

Realizamos una codificacion OneHotEncoding para las variables categoricas y un escalamiento de datos para las variables numericas.

Tambien realzamos un sobremuestreo con las muestras de "resultado = 1" en churn ya que teniamos un desbalance en los datos objetivo lo cual puede no ayudar a nuestros modelos a obtener un resultado satisfactorio.

Hacemos nuestra separacion de datos y realizamos el entrenamiento de diversos modelos como son: LogisticRegression, DecissionTreeClassifier, RandomForestClassifier, LGBMClassifier y CatboostClassifier; para los tres primeros casos utilizamos las herramientas de GridSearch y optuna.

Obtuvimos un mejor resultado AUC-ROC del modelo de LGBMClassifier con 0.836 aunado a que es el modelo mas rapido para desarrollar nuestra tarea con 0.20 segundos.